docs: block expressions (#5166)
dfinity/motoko: doc/md/fundamentals/4-control-flow/4-blocks.md

---
sidebar_position: 4
---

# Block expressions

> A block expression in Motoko is a sequence of statements enclosed in `{ ... }`. Blocks are used to group multiple operations, define local variables, and organize code for readability and logic.

Reviewer: ```suggestion
A block expression in Motoko is a sequence of declarations enclosed in `{ ... }`. 
Since every expression is also a declaration, the sequence can include expressions.
Blocks are used to group multiple operations, define local variables, and organize code for readability and logic.
```

> Block expressions are fundamental to function definitions, as they determine both the function’s structure and return value.

Reviewer: ```suggestion
Block expressions are fundamental to function definitions, as they determine both the function’s structure and return value.
```
```suggestion
Block expressions are typically used to define function and class bodies.
They can also be used as branches in conditional expressions and as the bodies of cases in switches and the arguments to `try-catch` expressions.

Since blocks  enclose declarations, the define new scopes for locally defined variables and types.
```

> Block expressions are fundamental to function definitions, as they determine both the function’s structure and return value.

Reviewer: ```suggestion
Block expressions are typically used to define function bodies and the bodies of `async` expressions.
They can also be used as branches in conditional expressions and as the bodies of cases in switches and the blocks of `try-catch-finally` expressions.
Since blocks  enclose declarations, they define new scopes for locally defined variables and types.
```

Each block contains a series of expressions and statements. The last evaluated expression determines the block’s result. If no meaningful final expression is present, the block returns `()`.

## Block expressions in functions

> Every function in Motoko contains a block expression that defines its behavior.

Reviewer: ```suggestion
Every function in Motoko contains a block expression that defines its body and behavior.
```

```motoko no-repl
// The function body is a block expression that returns the result of the last expression.
shared func add(x : Nat, y : Nat) : async Nat {
    let sum : Nat = x + y;

>     return sum;

Reviewer: ```suggestion
    sum;
```

};

// Example of using a block expression with conditional logic:
func classify(n : Int) : Text {
    if (n > 0) {
        "Positive"
    } else if (n < 0) {
        "Negative"
    } else {
        "Zero"
    } // The last evaluated branch determines the return value.
};
```

> ### Block assignments

Reviewer: ```suggestion
### `do` blocks
```

Blocks are not limited to functions; they can be used anywhere an expression is expected.

```motoko no-repl
let result : Nat = do {
  let x : Nat = 10;
  let y : Nat = 5;
  x * y // This value is returned and assigned to `result`
};
```

> ## `do` block

Reviewer: ```suggestion
```
Merge with previous section

The `do {}` expression in Motoko is used for executing multiple statements sequentially, while ensuring the block is treated as an expression. Unlike standard block expressions inside functions, the `do` block must return a value of type `()` which means it’s intended for side effects, not for returning meaningful values.

> If a statement inside the `do` block produces a value, that value can be discarded with the `ignore` keyword, unless it’s explicitly needed.

Reviewer: ```suggestion
```
Don't think we need to say this again here.

> 

Reviewer: A `do` block can also just return `()` and be evaluated for its side effect:
